crates/prune/prune/src/pruner.rs

```diff
@@ -169,81 +173,69 @@ impl<DB: Database> Pruner<DB> {
+                    }
+                }
+                ControlFlow::LimitReached => break,
             }
+        }
 
-            if let Some((to_block, prune_mode)) = segment
-                .mode()
-                .map(|mode| mode.prune_target_block(tip_block_number, segment.segment(), purpose))
-                .transpose()?
-                .flatten()
-            {
-                debug!(
-                    target: "pruner",
-                    segment = ?segment.segment(),
-                    ?purpose,
-                    %to_block,
-                    ?prune_mode,
-                    "Segment pruning started"
-                );
-
-                let segment_start = Instant::now();
-                let previous_checkpoint = provider.get_prune_checkpoint(segment.segment())?;
-                let output = segment.prune(
-                    provider,
-                    PruneInput { previous_checkpoint, to_block, limiter: limiter.clone() },
-                )?;
-                if let Some(checkpoint) = output.checkpoint {
-                    segment
-                        .save_checkpoint(provider, checkpoint.as_prune_checkpoint(prune_mode))?;
-                }
-                self.metrics
-                    .get_prune_segment_metrics(segment.segment())
-                    .duration_seconds
-                    .record(segment_start.elapsed());
-                if let Some(highest_pruned_block) =
-                    output.checkpoint.and_then(|checkpoint| checkpoint.block_number)
-                {
-                    self.metrics
-                        .get_prune_segment_metrics(segment.segment())
-                        .highest_pruned_block
-                        .set(highest_pruned_block as f64);
-                }
+        let initial_segments_len = self.segments.len();
+        let mut pruned_segments = Vec::with_capacity(self.segments.len());
+
+        // Prune other segments, taking them from the front of the queue
+        while let Some(segment) = self.segments.pop_front() {
+            let segment_name = segment.segment();
 
-                progress = output.progress;
-
-                debug!(
-                    target: "pruner",
-                    segment = ?segment.segment(),
-                    ?purpose,
-                    %to_block,
-                    ?prune_mode,
-                    %output.pruned,
-                    "Segment pruning finished"
-                );
-
-                if output.pruned > 0 {
-                    limiter.increment_deleted_entries_count_by(output.pruned);
-                    pruned += output.pruned;
-                    stats.insert(segment.segment(), (output.progress, output.pruned));
+            let result = self.prune_segment(
+                provider,
+                tip_block_number,
+                limiter,
+                &*segment,
+                PrunePurpose::User,
+            );
+            pruned_segments.push(segment);
```

Comment:
> a segment should only be added to pruned if it didn't quit because limit is reached

then we will prune the segment until completion, no? and it defeats the purpose of the ring

Reply:
it won't defeat its purpose, ring still ensures all segments are pruned till completion, not only same first few

another safe way would be to start with a random segment each prune job